voxel_grid/prototype/sfinae_check.hpp

```diff
@@ -0,0 +1,134 @@
+// #include <iostream>
+#include <optional>
+// #include <string>
+#include <type_traits>
+#include <vector>
+
+template <typename, typename>
+class TransformFilter;
+
+namespace detail {
+template <typename, typename>
+struct IsGridStruct;
+
+template <typename C, typename PointT>
+struct IsGridStruct {
+ private:
+  template <typename T>
+  static constexpr auto hasSetUp(T*) ->
+      typename std::is_same<decltype(std::declval<T>().setUp(
+                                std::declval<TransformFilter<T, PointT>*>())),
+                            bool>::type;
+  template <typename>
+  static constexpr std::false_type hasSetUp(...);
+
+  template <typename T>
+  static constexpr auto hasAddPoint(T*) -> typename std::is_same<
+      decltype(std::declval<T>().addPointToGrid(std::declval<PointT&>())),
+      void>::type;
+  template <typename>
+  static constexpr std::false_type hasAddPoint(...);
+
+  template <typename T>
+  static constexpr auto hasFilterBegin(T*) -> typename std::is_same<
+      decltype(std::declval<T>().filterGrid(
+          std::declval<decltype(std::declval<T>().begin())>())),
+      std::optional<PointT>>::type;
+  template <typename>
+  static constexpr std::false_type hasFilterBegin(...);
+
+  template <typename T>
+  static constexpr auto hasFilterEnd(T*) -> typename std::is_same<
+      decltype(std::declval<T>().filterGrid(
+          std::declval<decltype(std::declval<T>().end())>())),
```

Comment:
Instead of begin and end members of the class, can we test for validity of ADL with begin and end on the class?

Reply:
After trying for a while on it, the best way I found is to have one more function to do that

```cpp
template <typename T>
static constexpr auto hasFilterBegin(T*)-> typename std::is_same<decltype(std::declval<T>().filterGrid(begin(std::declval<T>()))),pcl::optional<PointT>>::type;
template <typename T>
static constexpr auto hasFilterBegin(T*)-> typename std::is_same<decltype(std::declval<T>().filterGrid(std::declval<T>().begin())),pcl::optional<PointT>>::type;
template <typename>
static constexpr std::false_type hasFilterBegin(...);
```

Reply:
Is the 2nd function needed? Would we have multiple definition error?

Reply:
> Is the 2nd function needed?

They can't be deduced by ADL after many trials at my side (with GCC 9)

> Would we have multiple definition error?

no error is thrown as long as not both `begin(std::declval<T>())` and `std::declval<T>().begin()` are defined. It's fine if ADL doesn't get triggered, I'll later check with other compilers with CI

Reply:
It doesn't throw error and does its job in CI, I think it's fine to put it into the PR